### PR TITLE
Use starkware/crypto-cpp for starknet verify/pedersen

### PR DESCRIPTION
**Pull request by utay**
Same idea as https://github.com/sorare/crypto/pull/2 but using https://github.com/starkware-libs/crypto-cpp which is audited and has existing JS bindings. 

## Problem

`sign`/`verify`/`pedersen` functions are CPU intensive and the current implementation uses https://github.com/indutny/elliptic which is very slow in JS.

## Solution

Use crypto-cpp for `pedersen` and `verify` functions that are called most of the time in our backend. `sign` is mostly called by client browsers and a `k` variable has to be computed which isn't trivial (https://github.com/starkware-libs/crypto-cpp/blob/master/src/starkware/crypto/ffi/js/crypto.js#L64-L65) so we decided to skip using it for now. 

crypto-cpp is audited and hasn't changed since Dec 2021 so I copy/pasted the C++ code directly in this repo. I've made a few changes in CMakeLists.txt and removed presubmit.sh and Go bindings as we don't use them. **To review this MR I suggest pulling the code and running `git diff origin/master ':!src/crypto-cpp'` to hide crypto-cpp additions.**   

The idea is to keep this library full JS compatible (for browsers especially) but set `USE_STARKWARE_CRYPTO_CPP=true` to make crypto functions use the C++ dynamic library and speed things up. 

## Results :zap: 

(on my machine, without C++ `sign`)

Before:
```
$ yarn test
yarn run v1.22.0
$ jest ./src
[...]
Done in 16.45s.
```

After:
```
$ USE_STARKWARE_CRYPTO_CPP=true yarn test
yarn run v1.22.0
$ jest ./src
[...]
Done in 2.71s.
```

### CODE REVIEW

File: src/starkware/signature.js [hunk 1]
```diff
@@ -96,6 +96,16 @@ function assertInRange(input, lowerBound, upperBound, inputName = '') {
...
 */
 export function pedersen(input) {
+  if (useCryptoCpp) {
+    if (typeof input[0] === 'string') {
+      input[0] = BigInt(`0x${input[0]}`);
+    }
+    if (typeof input[1] === 'string') {
+      input[1] = BigInt(`0x${input[1]}`);
+    }
+    return pedersenCpp(input[0], input[1]).toString(16);
```
Comment (utay) [inline src/starkware/signature.js:106]: To review :warning: Is it fine to assume that hexa strings will never start with `0x`? At least it's the case in tests

File: src/index.ts [hunk 2]
```diff
@@ -131,6 +133,15 @@ const sign = (privateKey: string, message: string): Signature => {
 };
 
 const verify = (publicKey: string, message: string, signature: Signature) => {
+  if (useCryptoCpp) {
+    return starkVerifyCpp(
+      BigInt(publicKey),
+      BigInt(`0x${message}`),
```
Comment (utay) [inline src/index.ts:139]: Same :warning: Is it fine to assume that hexa message strings will never start with `0x`? At least it's the case in tests

File: src/starkware/signature.js (hunk 1 above)
Comment (piedup) [reply src/starkware/signature.js:106]: Yes that's ok.

File: src/index.ts (hunk 2 above)
Comment (piedup) [reply src/index.ts:139]: Same.

**Review by piedup — APPROVED** [review]
Will I be able to install the lib on an M1?

**Comment by utay** [comment]
> Will I be able to install the lib on an M1?

@piedup As discussed we should make some tests about compiling the lib on M1/ARM, I'm not sure it will work as-is. https://github.com/starkware-libs/crypto-cpp/pull/4 is open on starkware's repo for this and can certainly help

**Pull request merged**